zipline/pipeline/loaders/quarter_estimates.py

```diff
@@ -0,0 +1,291 @@
+    merged = pd.merge(df1, df2, on='key')
+    return merged.drop('key', axis=1)
+
+
+class QuarterEstimatesLoader(PipelineLoader):
+    def __init__(self,
+                 estimates,
+                 base_column_name_map):
+        validate_column_specs(
+            estimates,
+            base_column_name_map
+        )
+
+        self.estimates = estimates[
+            estimates[EVENT_DATE_FIELD_NAME].notnull() &
+            estimates[FISCAL_QUARTER_FIELD_NAME].notnull() &
+            estimates[FISCAL_YEAR_FIELD_NAME].notnull()
+        ]
+
+        self.base_column_name_map = base_column_name_map
+
+    def load_quarters(self, num_quarters, dates_sids, final_releases_per_qtr):
+        pass
+
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        # TODO: how can we enforce that datasets have the num_quarters
+        # attribute, given that they're created dynamically?
+        groups = groupby(lambda x: x.dataset.num_quarters, columns)
+        out = {}
+        date_values = pd.DataFrame({SIMULTATION_DATES: dates})
+        # dates column must be of type datetime64[ns] in order for subsequent
+        # comparisons to work correctly.
+        date_values[SIMULTATION_DATES] = pd.to_datetime(date_values[
+            SIMULTATION_DATES
+        ])
+        estimates_all_dates = cross_product(date_values, self.estimates)
+        asset_df = pd.DataFrame({SID_FIELD_NAME: assets})
+        dates_sids = cross_product(date_values, asset_df)
+        for num_quarters in groups:
+            name_map = {c:
+                        self.base_column_name_map[
+                            getattr(c.dataset.__base__, c.name)
+                        ] for c in columns}
+
+            columns = groups[num_quarters]
+            # First, determine which estimates we would have known about on
+            # each date. Then, Sort by timestamp and group to find the latest
+            # estimate for each quarter.
+            final_releases_per_qtr = estimates_all_dates[
+                estimates_all_dates[TS_FIELD_NAME] <=
+                estimates_all_dates.dates
+            ].sort_values([TS_FIELD_NAME]).groupby(
+                [SIMULTATION_DATES,
+                 SID_FIELD_NAME,
+                 FISCAL_YEAR,
+                 FISCAL_QUARTER]
+            ).nth(-1).reset_index()
+
+            result = self.load_quarters(num_quarters,
+                                        dates_sids,
+                                        final_releases_per_qtr)
+
+            for c in columns:
+                column_name = name_map[c]
+                # Pivot to get a DataFrame with dates as the index and
+                # sids as the columns.
+                loader = DataFrameLoader(
+                    c,
+                    result.pivot(index=SIMULTATION_DATES,
+                                 columns=SID_FIELD_NAME,
+                                 values=column_name),
+                    adjustments=None
```

Comment:
Not yet. We don't apply them in the events loader either, so I'd imagine that we'd want to implement both in the future.